### PR TITLE
Attempt to implement the frame index box

### PR DESCRIPTION
**Pull request by jyrkialakuijala**
This PR tries to implement the "Part 2: File format" Frame Index Box,
aka 'jxli' box.

Likely it doesn't work yet correctly, but will allow trying to implement
a decoder for it.

### CODE REVIEW

File: lib/jxl/encode_internal.h [hunk 1]
```diff
@@ -20,6 +20,85 @@
...
+decoder are counted for this purpose, this excludes frames that are not intended
+for display but for compositing with other frames, such as frames that aren't
+the last frame with a duration of 0 ticks.
+
+
+    All the frames listed in jxli are keyframes and the first frame is
+    present in the list.
+    There shall be either zero or one Frame Index boxes in a JPEG XL file.
+    The offsets OFFi per frame are given as bytes in the codestream, not as
+bytes in the file format using the box structure. This means if JPEG XL Partial
```
Comment (lvandeve) [inline lib/jxl/encode_internal.h:46]: nit: the indentation of this comment doesn't match up on all lines

File: lib/jxl/encode.cc [hunk 2]
```diff
@@ -203,6 +203,82 @@ JxlEncoderStatus CheckValidBitdepth(uint32_t bits_per_sample,
...
+  static const int kVarintMaxLength = 10;
+  static const int kFrameIndexBoxHeaderLength = kVarintMaxLength + 8;
+  static const int kFrameIndexBoxElementLength = 3 * kVarintMaxLength;
+  const int buffer_size =
+      kFrameIndexBoxHeaderLength + NF * kFrameIndexBoxElementLength;
+  std::vector<uint8_t> buffer_vec(buffer_size);
+  uint8_t* buffer = buffer_vec.data();
+  size_t output_pos = 0;
+  ok &= jxl::EncodeVarInt(NF, buffer_vec.size(), &output_pos, buffer);
+  for (int i = 24; i >= 0; i -= 8) {
```
Comment (lvandeve) [inline lib/jxl/encode.cc:228]: nit: the function StoreLE32 from jxl/base/byte_order.h could be used for these

File: lib/jxl/encode.cc [hunk 3]
```diff
@@ -203,6 +203,82 @@ JxlEncoderStatus CheckValidBitdepth(uint32_t bits_per_sample,
...
+        // byte of previous frame from this index in the JPEG XL codestream. For
+        // the first frame, this is the offset from the first byte of the JPEG
+        // XL codestream.
+        OFFi -= frame_index_box.entries[prev_prev_ix].OFFi;
+      }
+      int32_t Ti = T_prev;
+      int32_t Fi = i - prev_ix;
+      ok &= jxl::EncodeVarInt(OFFi, buffer_vec.size(), &output_pos, buffer);
+      ok &= jxl::EncodeVarInt(Ti, buffer_vec.size(), &output_pos, buffer);
+      ok &= jxl::EncodeVarInt(Fi, buffer_vec.size(), &output_pos, buffer);
```
Comment (lvandeve) [inline lib/jxl/encode.cc:259]: in theory it could return false immediately here if !ok

File: lib/jxl/encode_internal.h (hunk 1 above)
Comment (jyrkialakuijala) [reply lib/jxl/encode_internal.h:46]: Heh, I had it beautiful and readable and then clang format made it like this. Trying to improve and see what will survive through clang format.

File: lib/jxl/encode.cc (hunk 2 above)
Comment (jyrkialakuijala) [reply lib/jxl/encode.cc:228]: Fixed

**Review by jonsneyers — APPROVED** [review]

**Review by sboukortt — APPROVED** [review]

**Pull request merged**